src/java.base/share/classes/java/util/zip/ZipInputStream.java

```diff
@@ -627,6 +628,70 @@ private void readEnd(ZipEntry e) throws IOException {
         }
     }
 
+    /**
+     * Returns true if the ZipEntry has a valid Zip64 extended information extra field.
+     *
+     * This method is used by {@link #readEnd(ZipEntry)} to determine whether the
+     * 'size' and 'compressed size' fields of the Data Descriptor record should
+     * be interpreted as 64-bit numbers instead of the regular 32-bit numbers.
+     *
+     * While ZipOutputStream only produces a Zip64 extra field if either the size or
+     * compressed size of the entry exceeds 0XFFFFFFFF, the ZIP APPNOTE.txt format
+     * specification explicitly allows the use of Zip64 format for entries of any size:
+     *
+     * <pre>
+     * When extracting, if the zip64 extended information extra
+     * field is present for the file the compressed and
+     * uncompressed sizes will be 8 byte values.
+     * </pre>
+     *
+     * To guard against invalid or corrupt extra fields, this method validates that
+     * any Zip64 extended field has a valid block size.
+     *
+     * This method returns false for any invalid extra block sizes, as if the extra
+     * data contained no Zip64 field.
+     */
+    private boolean hasZip64Extra(ZipEntry e) throws IOException {
+        byte[] extra = e.extra;
+        int fixedSize = 2 * Short.BYTES; // id + size
+        if (extra != null && extra.length > fixedSize) {
+            for (int i = 0; i < extra.length;) {
+                int id = get16(extra, i);
```

Comment:
This and other similar calls in this new method have a potential to throw an `ArrayIndexOutOfBoundsException` because we can't trust the `byte[]` returned by the `ZipEntry.getExtra()` call to be actually having the right amount of extra block data. The private `hasMagic` method in  `java.util.jar.JarOutputStream` has an example where we catch the `ArrayIndexOutOfBoundsException` when dealing with the extra data, we should do similar here and return `false`.

(More about the trustworthiness of that byte[] value as a separate comment)